x/incentive/keeper/refundable_msg_index.go

```diff
@@ -1,31 +1,99 @@
 
-// getRecipient returns the address that should receive the refund.
-func getRecipient(feeTx sdk.FeeTx) sdk.AccAddress {
-	if feeGranter := feeTx.FeeGranter(); feeGranter != nil {
-		return feeGranter
-	}
-	return feeTx.FeePayer()
-}
+	"github.com/babylonlabs-io/babylon/v4/x/incentive/types"
+)
 
-// RefundTx refunds the given tx by sending the fee back to the fee payer.
+// RefundTx refunds the transaction fee to the appropriate party.
+// If a fee grant was used, it restores the granted allowance.
+// Otherwise, it sends the fee back to the original fee payer.
 func (k Keeper) RefundTx(ctx context.Context, tx sdk.FeeTx) error {
 	txFee := tx.GetFee()
 	if txFee.IsZero() {
-		// not possible with the global min gas price mechanism
-		// but having this check for compatibility in the future
 		return nil
 	}
-	txFeePayer := getRecipient(tx)
 
-	return k.bankKeeper.SendCoinsFromModuleToAccount(ctx, k.feeCollectorName, txFeePayer, txFee)
+	feeGranter := tx.FeeGranter()
+	feePayer := tx.FeePayer()
+
+	if feeGranter != nil && !bytes.Equal(feeGranter, feePayer) {
+		return k.refundToFeeGranter(ctx, feeGranter, feePayer, txFee)
+	}
+
+	// refund to fee payer
+	return k.bankKeeper.SendCoinsFromModuleToAccount(ctx, k.feeCollectorName, feePayer, txFee)
+}
+
+// refundToFeeGranter restores the fee grant allowance by increasing the spend limit or restoring the deleted grant
+// and sends the refund to the fee granter
+func (k Keeper) refundToFeeGranter(ctx context.Context, feeGranter, feePayer sdk.AccAddress, refund sdk.Coins) error {
+	allowance, err := k.feegrantKeeper.GetAllowance(ctx, feeGranter, feePayer)
+
+	if err != nil {
+		if errorsmod.IsOf(err, sdkerrors.ErrNotFound) || errorsmod.IsOf(err, feegrant.ErrNoAllowance) {
+			// Allowance was totally depleted and deleted
+			// Need to restore it
+			err := k.restoreDeletedFeeGrant(ctx, feeGranter, feePayer, refund)
+			if err != nil {
+				return err
+			}
+		} else {
+			// got another error
+			return err
+		}
+		// restored fee grant,
+		// send refund to feeGranter
+		return k.bankKeeper.SendCoinsFromModuleToAccount(ctx, k.feeCollectorName, feeGranter, refund)
+	}
+
+	// Existing allowance still present — just increase spend limit
+	updatedAllowance, err := restoreAllowanceSpent(allowance, refund)
+	if err != nil {
+		return err
+	}
+
+	if err := k.feegrantKeeper.UpdateAllowance(ctx, feeGranter, feePayer, updatedAllowance); err != nil {
+		return fmt.Errorf("failed to update fee allowance: %w", err)
+	}
+
+	return k.bankKeeper.SendCoinsFromModuleToAccount(ctx, k.feeCollectorName, feeGranter, refund)
+}
+
+// restoreDeletedFeeGrant restores the fee grant as a basic allowance
+// with a 2 day expiration
+func (k Keeper) restoreDeletedFeeGrant(ctx context.Context, feeGranter, feePayer sdk.AccAddress, refund sdk.Coins) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	// Set expiration time (48 hours from current block time)
+	expiration := sdkCtx.BlockTime().Add(48 * time.Hour)
```

Comment:
why do we keep the 2 days?

Reply:
Good question! Here we have to decide what to do, if set a hardcoded expiration time (could be 2 days or whatever we decide), or somehow try to get what was the consumed grant expiration. For the time being, I believe setting a hardcoded expiration time is the best approach. Then we have to decide what is the period we want to set here

Reply:
refactored the code to not restore the depleted allowance for this case.
So we avoid setting up a different allowance type than the one depleted